internal/kargo/promotion_builder.go

```diff
@@ -0,0 +1,259 @@
+package kargo
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/oklog/ulid/v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/api/user"
+)
+
+const (
+	// PromotionAliasSeparator is the separator used in the Promotion step alias
+	// to separate the task alias from the step alias.
+	PromotionAliasSeparator = "::"
+
+	// nameSeparator is the separator used in the Promotion name.
+	nameSeparator = "."
+
+	// ulidLength is the length of the ULID string.
+	ulidLength = ulid.EncodedSize
+
+	// shortHashLength is the length of the short hash.
+	shortHashLength = 7
+
+	// maxStageNamePrefixLength is the maximum length of the Stage name
+	// used in the Promotion name prefix before it exceeds the Kubernetes
+	// resource name limit of 253.
+	maxStageNamePrefixLength = 253 - len(nameSeparator) - ulidLength - len(nameSeparator) - shortHashLength
+)
+
+type PromotionBuilder struct {
+	client client.Client
+}
+
+// NewPromotionBuilder creates a new PromotionBuilder with the given client.
+func NewPromotionBuilder(c client.Client) *PromotionBuilder {
+	return &PromotionBuilder{
+		client: c,
+	}
+}
+
+// Build creates a new Promotion for the Freight based on the PromotionTemplate
+// of the given Stage.
+func (b *PromotionBuilder) Build(
+	ctx context.Context,
+	stage kargoapi.Stage,
+	freight string,
+) (*kargoapi.Promotion, error) {
+	if stage.Name == "" {
+		return nil, fmt.Errorf("stage is required")
+	}
+
+	if stage.Spec.PromotionTemplate == nil {
+		return nil, fmt.Errorf("stage %q has no promotion template", stage.Name)
+	}
+
+	if freight == "" {
+		return nil, fmt.Errorf("freight is required")
+	}
+
+	// Build metadata
+	annotations := make(map[string]string)
+	if u, ok := user.InfoFromContext(ctx); ok {
+		annotations[kargoapi.AnnotationKeyCreateActor] = kargoapi.FormatEventUserActor(u)
+	}
+
+	promotion := kargoapi.Promotion{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        generatePromotionName(stage.Name, freight),
+			Namespace:   stage.Namespace,
+			Annotations: annotations,
+		},
+		Spec: kargoapi.PromotionSpec{
+			Stage:   stage.Name,
+			Freight: freight,
+			Vars:    stage.Spec.PromotionTemplate.Spec.Vars,
+			Steps:   stage.Spec.PromotionTemplate.Spec.Steps,
+		},
+	}
+	return &promotion, nil
+}
+
+// InflateSteps inflates the Promotion steps by resolving any references to
+// PromotionTasks and expanding them into their individual steps. The inflated
+// steps are then set on the Promotion, replacing the original steps.
+func (b *PromotionBuilder) InflateSteps(ctx context.Context, promo *kargoapi.Promotion) error {
```

Comment:
Making sure I understand the rational for having InflateSteps() separate from Build()...

I presume that by not having Build() automatically inflate, we are preserving the possibility of a Promo having been defined manually and submitted via kubectl, in which case, building isn't necessary, but expansion is?

Reply:
`Build` is for creating a Promotion based on a Stage (or rather, the Promotion template within). Which we do in several places, either through the API or CLI.

`InflateSteps` is what always happens on the server-side, and ensures that even if a Promotion would be created through `kubectl` (as you suggest) — we still handle the core logic of correctly inflating any references we find into actual steps.

Reply:
Ok... so I agree with the rationale for keeping them separate, but is there any reason we inflate server side, but we don't build server side?

Could the mutating webhook we discussed build if steps == nil and then inflate unconditionally? As opposed to calling build in several different places?

Reply:
We discussed this offline, and partly because of debugging reasons and introducing too many sprinkles of magic. It is better to keep the builder in place so you have some (meaningful) YAML to hold on to.